Add examples/fetch.rs to fetch spdx.org/licenses and format it as src/spdx.rs

nodakai opened the pull request:
Run `cargo run --example fetch-license-list-from-spdx > spdx.rs` and rename it as `src/spdx.rs`.

This tool will make manual processes like #10 unnecessary.

It was written as an "example" as a trick to circumvent Cargo's limitation that dependencies cannot be added to a subset of project outputs (`hyper` *etc.* were added as `dev-dependencies`.)

nodakai committed:
Add examples/fetch.rs to fetch spdx.org/licenses and format it as src/spdx.rs

Signed-off-by: NODA, Kai <nodakai@gmail.com>

wking commented on examples/fetch.rs:
```diff
@@ -0,0 +1,175 @@
+
+    writeln!(stdout, "\
+/*
+ * whitelist fetched from http://spdx.org/licenses/
```
This comment seems stale, since below you are using license-list-data's JSON.

wking commented on examples/fetch.rs:
```diff
@@ -0,0 +1,175 @@
+
+    let mut core = Core::new()?;
+
+    download(&mut core, "https://raw.githubusercontent.com/spdx/license-list-data/master/json/licenses.json", |json| {
```
Instead of floating with master, I think it's better to take a tag argument and use that here.  For example:

```console
cargo run --example fetch-license-list-from-spdx v3.0
```

would pull from https://raw.githubusercontent.com/spdx/license-list-data/v3.0/json/licenses.json

nodakai replied:
Git tag is now supported (sorry for the delay!)

nodakai committed:
examples/fetch.rs: support a Git tag in command line

Run

    cargo run --example fetch-license-list-from-spdx v3.0 > spdx.rs

to download from https://github.com/spdx/license-list-data/tree/v3.0/json

Signed-off-by: NODA, Kai <nodakai@gmail.com>

wking commented on Cargo.toml:
```diff
@@ -1,7 +1,20 @@
 authors = ["Without Boats <woboats@gmail.com>"]
-license = "Apache-2.0 OR MIT"
-description = "Validate SPDX 2.0 license expressions."
+license = "Apache-2.0/MIT"
```
This is reverting #16; can you leave the `OR` form as it was?

nodakai replied:
Fixed accordingly

wking commented on Cargo.toml:
```diff
@@ -1,7 +1,20 @@
 [package]
 name = "license-exprs"
-version = "1.3.0"
+version = "1.4.0"
```
We'll want to bump the version before we push a new release (hopefully soon), but I'd rather have version bumps in their own pull requests, not mixed in with logic changes.  Can you leave this at 1.3.0 for this PR?

wking reviewed (commented):
Two minor nits, otherwise this looks good to me :).

nodakai committed:
Switch from Hyper to Reqwest

Also improve JSON validation

No change to the output file

Signed-off-by: NODA, Kai <nodakai@gmail.com>

wking commented on examples/fetch.rs:
```diff
@@ -0,0 +1,194 @@
+            "-d" => {
+                debug = true;
+            },
+            s if s.starts_with('v') && s.ends_with(|c: char| c.is_ascii_digit()) => {
```
Can we drop `&& s.ends_with(|c: char| c.is_ascii_digit())` here?  It's breaking `cargo build --examples` on my 1.23.0-dev `rustc` with:

```
error: use of unstable library feature 'ascii_ctype' (see issue #39658)
  --> examples/fetch.rs:62:64
   |
62 |             s if s.starts_with('v') && s.ends_with(|c: char| c.is_ascii_digit()) => {
   |                                                                ^^^^^^^^^^^^^^
   |
   = help: add #![feature(ascii_ctype)] to the crate attributes to enable

error: aborting due to previous error

error: Could not compile `license-exprs`.
```

And I think matching on the opening `v` is a sufficient indicator of version-ness, even without the trailing-digit check.

wking commented:
Also, there's no need for `Signed-off-by` in the commit messages.  I don't think those are meaningful without repo-local documentation for what's being signed (like [this][1] or [this][2]).  I don't have a problem with you keeping them in your commit messages anyway, but don't feel like you need to have them there on my account ;).

[1]: https://github.com/git/git/blob/v2.17.0/Documentation/SubmittingPatches#L278
[2]: https://github.com/wking/signed-off-by

nodakai commented:
@wking You have an interesting project!
